### ISSUE TITLE
Bright flash of BFG causing headaches

### ISSUE THREAD

**Issue by Parker1200** [rdntcntrl/ratoa_gamecode#41]
In Ratmod, after firing the BFG, a very bright flash appears on the screen and lasts about a frame. This doesn't happen in vanilla OA or when you set g_delagMissiles to 0.  This appears to be caused by the delagged projectiles simply becoming visible too early, resulting in flashing. This only affects the BFG, other weapons do not have this problem. This is a significant problem that affects health and limits the enjoyment of playing Ratmod. Some people experience headaches and migraines very quickly when playing in "BFG only" mode or on maps with a lot of BFG. I'm not sure to what extent people with epilepsy might be affected by this.

I would suggest adding new graphics for the BFG projectile, just like we have for rockets. Maybe even different colors besides green, like red and purple (also explosions). Maybe the delag code can also be tweaked to eliminate the bright screen flash when using regular OpenArena graphics?

**Comment by Parker1200** [comment]
I created a patch that basically does exactly what I wanted:
https://github.com/Parker1200/ratoa_gamecode/commit/9ab704784e6c3dcd4d5505cd6c8b961c4e3bf696

Delagged BFG projectiles no longer appear directly in front of the player, but they do after a certain amount of time, so it looks like in vanilla OA. I tuned the time before the projectile should appear using a local server and then tested it online.

I still think that new graphics for BFG projectiles in different color variants would be a cool addition. BFG in Q3 with cg_drawGun set to 0 was easier on the eyes, so it can be done better. The projectiles had a tail, which is missing in OA.

**Comment by rdntcntrl** [comment]
I think the flash you're seeing is caused by the shader which makes the projectile's tail (yes it does have one) pulse. Just disabling that seems to be enough.

scripts/weapon_bfg.shader:
```
BFGTrail
{
        nopicmip
        cull disable
        {
                clampmap models/weaphits/bfg2.tga
                blendfunc add
                rgbGen const ( 0.694118 0.694118 0.694118 )
                tcMod rotate 1522
        }
        {
                clampmap models/weaphits/bfg2.tga
                blendfunc add
                rgbGen const ( 0.286275 0.54902 0.25098 )
                tcMod rotate -1222
                
                // This line caused the flash
                //tcMod stretch sin 1.3 0.3 0 4
        }
}
```

You could test if this is the cause for what you're seeing by enabling vertex light (`r_vertexlight 1`), which among other things disables that shader stage.

It would probably also help to just make the projectile smaller, like we did the rockets.

> I created a patch that basically does exactly what I wanted: [Parker1200@9ab7047](https://github.com/Parker1200/ratoa_gamecode/commit/9ab704784e6c3dcd4d5505cd6c8b961c4e3bf696)

I think your patch would also break the predicted explosions if they happen within that time. Although if fixing the visuals is enough I'd prefer not to add any workarounds at all. 

You can also completely disable locally predicted missiles on the client side using `cg_ratPredictMissiles 0`. Although adding an option to not entirely disable it but just delay showing missiles (for all types) by a configurable amount of time is perhaps something I have think about.

**Comment by Parker1200** [comment]
I can confirm that modifying the shader fixes the BFG flash, and also `r_vertexlight 1` removes it without any modifications.

There is still one issue that persists even in the base game and is not related to delagging in any way. When bots are followed in spectator mode, a large BFG flash is visible. However, this is also present in base OA and even base Q3A. I even checked Devotion (partial Ratmod conversion to Q3A) and apparently it inherited the BFG flashing issue due to delagging. To summarize:

| game | action | flash |
| --- | --- | --- |
| baseoa | playing| no flash |
| baseoa |following a bot | flash present |
| Ratmod | playing | flash present before the shader was modified |
| Ratmod | following a bot | flash present |
| baseq3 | playing | no flash |
| baseq3 | following a bot | flash present |
| Devotion | playing | flash present (very bad, although I might like the projectile design better overall) |
| Devotion | following a bot |flash present |

Basically, delagging causes the same problem in both games and the flash in spectator mode would be difficult to remove since it has always been present so far. (I didn't check actual players, only bots.)

I modified my workaround:
https://github.com/Parker1200/ratoa_gamecode/commit/9334dfe75796b51de1a6b5c474ee9f0af9ddc19d

Even though this is no longer necessary thanks to the shader tweak, I must admit that I like the combined effect of both fixes. With workaround active, the initial size of the fired projectile is practically the same as in vanilla OA, and thanks to the new shader you can't get the flash effect when moving forward very fast. Perhaps a smaller variant of the projectile would completely invalidate the rationale for this workaround.

**Comment by rdntcntrl** [comment]
> Even though this is no longer necessary thanks to the shader tweak, I must admit that I like the combined effect of both fixes. With workaround active, the initial size of the fired projectile is practically the same as in vanilla OA, and thanks to the new shader you can't get the flash effect when moving forward very fast. Perhaps a smaller variant of the projectile would completely invalidate the rationale for this workaround.

We could also increase the default `g_delagMissileBaseNudge`, which will move the projectile forward (not just the visuals). This would also have an effect when spectating bots. IIRC in baseoa it gets moved by 50ms by default, whereas in Ratmod the default is only 10ms. The reason for the lower value is mainly to give players a better chance to dodge projectiles despite the delag (many really dislike getting hit by "invisible rockets").

Perhaps a value of 20-25ms would be more appropriate though and together with the fixed shader might be enough to make the visuals a bit better.
I agree that the BFG effects in general (sprites and the projectile model/size) could be improved, but that's probably not going to happen anytime soon.

**Comment by Parker1200** [comment]
For now, the BFG flash issue is fixed well enough. It is now possible to play for longer without feeling any discomfort, at least in my opinion. I don't think increasing `g_delagMissileBaseNudge` would be a good thing, especially because of one of the weapons or spectator mode. Due to changes in Ratmod v0.17.0, I will even have to do the opposite. Now the missiles behave as if `g_delagMissileBaseNudge` has already been increased, so I will have to lower the value of this cvar even to `0` to recreate the correct feel of the entire game. Zero behaves like the old `10` in my tests. I hope that there will be no side effects of reducing this value to zero. The current delagging changes are especially noticeable on weapons that have higher cooldowns and fast projectiles, such as rockets with a speed set to `1800`.

**Comment by rdntcntrl** [comment]
> Due to changes in Ratmod v0.17.0, I will even have to do the opposite. Now the missiles behave as if `g_delagMissileBaseNudge` has already been increased, so I will have to lower the value of this cvar even to `0` to recreate the correct feel of the entire game. Zero behaves like the old `10` in my tests. I hope that there will be no side effects of reducing this value to zero. The current delagging changes are especially noticeable on weapons that have higher cooldowns and fast projectiles, such as rockets with a speed set to `1800`.

Changing it to 0 shouldn't give you any weird side effects, and especially if you set the rocket speed to 1800 it's perfectly fine to do that IMO.

But I'm curious as to why you're seeing such a difference with v0.17.0. It really shouldn't result in any drastic changes (I can't see any noticeable difference visually, even with 1800 rockets).
Could you share some of the settings you use, particularly `sv_fps` and all the `g_delagMissile*` cvars on the server side and `snaps`, `rate`, `cl_timenudge` and `com_maxfps` on the client side?

**Comment by Parker1200** [comment]
Sure, here's my config:

Server side:
```
sv_fps 40
g_delagMissileDebug = "0"
g_delagMissileBaseNudge = "10"
g_delagMissileMaxLatency = "500"
g_delagMissiles = "1"
g_delagMissileLimitVariance = "1"
g_delagMissileLimitVarianceMs = "25"
g_delagMissileLatencyMode = "1"
g_delagMissileCorrectFrameOffset = "1"
g_delagMissileNudgeOnly = "0"
g_delagMissileImmediateRun = "2"
```

Client side:
```
snaps 40
rate 100000 (/status on the server shows it is 90000)
cl_timenudge 0
com_max_fps 250
```

I see quite a difference when I test online and when I shoot rockets through a portal. Exactly the same configuration, v0.17.0 the rockets disappear much faster as I get closer to the portal. In v0.16.3 I can get closer to the portal and they will still appear. I also notice a difference between the two versions when playing casually.

Also the BFG projectile materializes further away from the player with the new version, so it appears smaller. It becomes visually "larger", as it used to be in the older version, after lowering the value of `g_delagMissileBaseNudge`. (all without using my workaround)

**Comment by rdntcntrl** [comment]
Thanks. Looks like the defaults, which is good.

> I see quite a difference when I test online and when I shoot rockets through a portal. Exactly the same configuration, v0.17.0 the rockets disappear much faster as I get closer to the portal. In v0.16.3 I can get closer to the portal and they will still appear. I also notice a difference between the two versions when playing casually.

Strange, I cannot reproduce this. Testing with plasmagun and rockets on aggressor, the distance from the portal (checked with `\where`) at which I can still see the projectile before it disappears seems to be the same for me in v0.17.0 as it is in v0.16.3.

> Also the BFG projectile materializes further away from the player with the new version, so it appears smaller. It becomes visually "larger", as it used to be in the older version, after lowering the value of `g_delagMissileBaseNudge`. (all without using my workaround)

Keep in mind that I did make the BFG model quite a bit smaller.

**Comment by Parker1200** [comment]
Are you absolutely sure that there is not the slightest difference? If this is the case, then there is some other factor at play than just the Ratmod version. I have no idea, maybe my hardware or something else?

I repeated the tests with rockets (rocket speed: `1800`) and got the same results using the `/where` command. This time I reverted the commit that makes the difference for me and tested it locally. I also included one more server that is publicly available online. My test map was estatica. I used the portal that divides the slightly longer corridor on the side. Every time I moved, I used the center view button. I wrote down the coordinates only when the rockets were clearly visible after firing each time. The coordinate in the middle decreases as I approach the portal:

| version | server | g_delagMissileBaseNudge | coordinates |
| --- | --- | --- | --- |
| v0.17.0 no modifications | localhost | 10 | -570 -3 56 |
| v0.17.0 https://github.com/rdntcntrl/ratoa_gamecode/commit/2d6912a9b974231e6a9e9439a6fa6d6d7746d034 reverted | localhost | 10 | -571 -25 56 |
| v0.17.0 no modifications | localhost | 0 | -570 -22 56 |
| v0.17.0 | /N/City DE | 10 | -569 -3 56 |

> Keep in mind that I did make the BFG model quite a bit smaller.

Sorry, I didn't notice that there was another commit, this time modifying the md3 model. Thanks for the fix and pointing it out.

I think I will set `g_delagMissileBaseNudge` to `0` on both of my servers after the upgrade. One of them has standard weapons, so I'll test it out before I make a decision.

**Comment by rdntcntrl** [comment]
What is your screen's refresh rate? Are you getting stable FPS (FPS counter not flickering)?

The visual test has limitations. I'm currently on a 60hz screen, so for a projectile with a velocity of 1800 I would expect the distance at which it is still visible to vary by about `1800 * 1/60 = 30` depending on when exactly it is fired. So if you go by visuals it's important to take multiple samples (and also to not keep pressing `+attack`, but instead releasing it after every shot so you get a fresh timing).
I also prefer testing it with PG as there aren't multiple models (it's just a sprite) and the speed is not configurable, reducing the chance of making some mistake. PG speed is 2000, so `2000 * 1/60 = 33`.

Keeping this in mind, I tried your test (with 1800 rockets) and got about the same average distance for both versions (usually around -20 to -10).

I've made a test branch 341511b, which has the delag commit reversed but adds debug output to show the launch time difference between the two versions. Positive numbers mean the old version would show the projectile closer to the player, negative numbers the opposite.
I took 50 samples at 250fps (again, releasing `+attack` every time) and got an average difference of only `-0.14ms`. The biggest differences I got were 3 and -2. For PG, that would equate a distance of only 6.

I don't see a problem with you setting `g_delagMissileBaseNudge` to `0` (just don't make it negative) if that's what you prefer,  I'm just trying to make sure the delag code is not actually messed up.

**Comment by Parker1200** [comment]
My refresh rate is also 60Hz and I play on low-end hardware (Mini PC), so the framerate is not entirely stable. I can make it stable with `/com_busyWait 1`, but sometimes I just need to lower the maximum framerate... However, thanks to your delag debug branch, I now know what's going on. I can confirm that everything was fine with the new delagging. The only problem was with my testing and focusing on visuals. I also keep getting low spreads in values like `-2` and `2`. There was a larger spread in the `-3` and `6` values in one game, but that only happened once. (it must have just been a freak accident)

I also now have a theory as to why I got such skewed results during visual testing. There was one thing I didn't mention because I did the testing correctly, but it still affected my perception in some way. Since I play on a weak computer, I added a frameskip feature to my client to have better physics in this game without having to buy new hardware. During testing I turned it off because I knew it might affect the results (I took screenshots so I know it was turned off when I saved the coordinates), but during testing I turned frame skipping on and off because I was curious what effect it might have in my isolated case... and somehow I managed to wrongly convince myself after testing there wasn't even a real difference between the two settings. These projectiles were sometimes able to disappear closer to the portal, even though I had frameskip turned off. So my visual tests were very unreliable. One thing I knew for sure was that something had changed in the game after the last Ratmod update.

Now, thanks to your delag debug code, I can forget about looking at visuals and see what the problem was with my casual gameplay. When I enable frameskip and set it to `2` (every second frame will be skipped, `com_maxfps 250` becomes 125 fps) I get values like `4` and `5`. If I set frameskip to `4` (`com_maxfps 250` becomes 62 fps) it gets much worse as values like `10` and `12` appear in the debug output.

Here's how my frameskip works and how I "fixed it":

Inside the `void CL_Frame ( int msec )` function I just count the frames and skip every second call to `SCR_UpdateScreen();` or more if my `cl_frameSkip` cvar value is set to more. This always worked fine, but after the Ratmod update it broke. Now I can't just skip `SCR_UpdateScreen();` and I have to call a function inside `cgvm` and execute the code below to make the new delagging work properly with my very simple frameskip:

```
cg.time = serverTime;
cg.frametime = cg.time - cg.oldTime;
if ( cg.frametime < 0 ) {
    cg.frametime = 0;
}
cg.oldTime = cg.time;
```

This is a huge deal for me (well, only for me since no one else is using ioq3 with this patch) because I need better physics to play "only rockets". For example, there is a map on the City servers called "w4warehouse" where I can jump to the roof of a building directly from the ground when `com_maxFPS` is set to `250` or higher. This can completely change the outcome of the game... By skipping frames, I can compete with players who have better hardware specs. It would be nice if all players had the same physics, but it would have to be the "better physics" with higher jumps, not some broken fixes that would make the game feel wrong for all players, especially skilled ones. I also wouldn't like to see any changes to `g_knockback` or gravity on the City servers. Everything is perfect for me regarding the physics of the game. The only problem is the lack of consistency. If only someone could figure out how to actually make the physics work for all players as if it were 333 frames per second. By the way, my servers are set to `pmove_float 1` and `pmove_fixed 0`.

I could revert [the commit](https://github.com/rdntcntrl/ratoa_gamecode/commit/2d6912a9b974231e6a9e9439a6fa6d6d7746d034) in my Ratmod fork (used on 2 servers), but then I would have to do more maintenance and I admit I don't understand delagging well. I will have to use the code above instead. This will not affect other players and will allow me to play normally again (with frameskip), although only on my servers. Of course, if someone else had the same idea of using frame skipping, their client would also need a patch to adapt to the new delagging. I know there are some very exotic clients and I don't know all their features.

Sorry for the confusion regarding delagging!

**Comment by rdntcntrl** [comment]
> My refresh rate is also 60Hz and I play on low-end hardware (Mini PC), so the framerate is not entirely stable. I can make it stable with `/com_busyWait 1`, but sometimes I just need to lower the maximum framerate...

If `com_busyWait 1` stabilizes your framerate, it may be that CPU frequency scaling is to blame for the unstable FPS. If the frequency is reduced while the CPU is idle it can throw the frame timing off. Instead of using busywait, which just wastes CPU cycles, try disabling frequency scaling (or setting it to be less aggressive) while playing. On Linux you can use `cpufrequtils` (`cpufreq-set`).
I've had this happen on both on slow (performance limited by graphics card) and fast computers.

> The only problem is the lack of consistency. If only someone could figure out how to actually make the physics work for all players as if it were 333 frames per second. By the way, my servers are set to `pmove_float 1` and `pmove_fixed 0`.

Interesting. `pmove_float 1` should really make the physics framerate-independent, but I know it's not perfect (for instance when you get pushed by damage).

But if you really want exactly the same physics, why not just set `pmove_fixed 1`? E.g. if you want 333fps physics:
```
set pmove_float 0
set pmove_fixed 1
set pmove_msec 3
```

> Sorry for the confusion regarding delagging!

No problem, it's always good to test these kinds of changes thoroughly anyway!
